### PR TITLE
Sort persistent notifications ascending

### LINKED ISSUE #6565: Change sort order of Persistent Notifications (show most recent notifications at top)
## The request

This one is fairly self explanatory. Currently new persistent notifications appear at the bottom of the list, while literally every other operating system and website/app/service that I'm aware of sorts notifications by most recent at top. 

Here is an example image to demonstrate. 

image

With just two notifications this is not too bad, but once you start getting more than that your most important/recent stuff is behind a scroll wheel. I use the [Feedreader integration](https://www.home-assistant.io/integrations/feedreader/) to display a few RSS feeds here and I don't dismiss them until I get around to reading the articles. But this often causes me to not see the most recent notifications, because they're buried at the very bottom often behind a scroll.

I think it should be sorted in the way that people expect, consistent with iOS, Android, macOS, etc.

### PR DESCRIPTION
**Pull request by Misiu**
## Proposed change

sort persistent notifications, to show the most recent on top.

The sort is done in code, the other option is to use https://polymer-library.polymer-project.org/3.0/api/elements/dom-repeat#DomRepeat-property-sort, but notification-drawer will be rewritten someday to Lit, so the first option was easier.

## Type of change

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (thank you!)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example configuration

```yaml

```

## Additional information

- This PR fixes or closes issue: fixes #6565
- This PR is related to issue:
- Link to documentation pull request:

## Checklist

- [x] The code change is tested and works locally.
- [x] There is no commented out code in this PR.
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

[docs-repository]: https://github.com/home-assistant/home-assistant.io

### CODE REVIEW

File: src/dialogs/notifications/notification-drawer.js [hunk 1]
```diff
@@ -139,7 +139,15 @@ export class HuiNotificationDrawer extends EventsMixin(
       .filter((entityId) => computeDomain(entityId) === "configurator")
       .map((entityId) => hass.states[entityId]);
 
-    return notificationsBackend.concat(configuratorEntities);
+    const notifications = notificationsBackend.concat(configuratorEntities);
+
+    notifications.sort(function (n1, n2) {
+      if (n1.created_at > n2.created_at) return -1;
+      if (n1.created_at < n2.created_at) return 1;
```
Comment (bramkragten) [inline src/dialogs/notifications/notification-drawer.js:146]:
```suggestion
      if (n1.created_at > n2.created_at) { return -1; }
      if (n1.created_at < n2.created_at) { return 1; }
```

**Review by bramkragten — APPROVED** [review]

File: src/dialogs/notifications/notification-drawer.js [hunk 2]
```diff
@@ -139,7 +139,15 @@ export class HuiNotificationDrawer extends EventsMixin(
       .filter((entityId) => computeDomain(entityId) === "configurator")
       .map((entityId) => hass.states[entityId]);
 
-    return notificationsBackend.concat(configuratorEntities);
+    const notifications = notificationsBackend.concat(configuratorEntities);
+
+    notifications.sort(function (n1, n2) {
+      if (n1.created_at > n2.created_at) { return -1; }
```
Comment (bramkragten) [inline src/dialogs/notifications/notification-drawer.js:145]: btw, what is created_at? An iso date string probably? You can't compare that, you should parse it as a date fist.

Comment (Misiu) [reply src/dialogs/notifications/notification-drawer.js:145]: Ok, will do in a second.

Comment (Misiu) [reply src/dialogs/notifications/notification-drawer.js:145]: done

File: src/dialogs/notifications/notification-drawer.js [hunk 3]
```diff
@@ -139,7 +139,19 @@ export class HuiNotificationDrawer extends EventsMixin(
...
       .map((entityId) => hass.states[entityId]);
 
-    return notificationsBackend.concat(configuratorEntities);
+    const notifications = notificationsBackend.concat(configuratorEntities);
+
+    notifications.sort(function (n1, n2) {
+      if (new Date(n1.created_at) > new Date(n2.created_at)) {
+        return -1;
+      }
+      if (new Date(n1.created_at) < new Date(n2.created_at)) {
```
Comment (bramkragten) [inline src/dialogs/notifications/notification-drawer.js:148]: Don't do the date parsing twice, store it in a variable.

Comment (Misiu) [reply src/dialogs/notifications/notification-drawer.js]: done

**Review by bramkragten — APPROVED** [review]

**Comment by Misiu** [comment]
@bramkragten should I add undefined check as you suggested here: https://github.com/home-assistant/frontend/pull/7199#discussion_r498826335

**Comment by bramkragten** [comment]
> Should I add undefined check as you suggested here: [#7199 (comment)](https://github.com/home-assistant/frontend/pull/7199#discussion_r498826335)

Depends on the type, can it be undefined?

**Comment by ludeeus** [comment]
Instead of the logic added here, and more or less the same logic in #7199, shouldn't we rather make a new helper fo compareDatetime?

**Comment by Misiu** [comment]
> Depends on the type, can it be undefined?

According to the interface, it can't be undefined.
Ref:
```
export interface PersistentNotification {
  created_at: string;
  message: string;
  notification_id: string;
  title: string;
  status: "read" | "unread";
}
```

**Comment by Misiu** [comment]
@bramkragten should I do something else in this PR?
My compare logic and the one mentioned by @ludeeus are a bit different:
https://github.com/home-assistant/frontend/pull/7199/files#diff-17f747d00c7a961b75947a7ba8215710837fa3a9ea39a1e5886e16ab5363866bR27
https://github.com/home-assistant/frontend/pull/7195/files#diff-c2a273c22fe408b3d6d49a7ab8befabf96f7e1c74988567902ca79336d179471R144

so really no ideas on how to make this a generic function.

**Comment by Misiu** [comment]
Something like this: https://stackblitz.com/edit/typescript-yzualr?

```ts
function SortByDateAscending<T, K extends keyof T>(values: T[], key: K) {
  const compare = (first: T, second: T) => {
    const timeA = first[key] ? new Date(first[key]) : 0;
    const timeB = second[key] ? new Date(second[key]) : 0;
    if (timeA < timeB) {
      return 1;
    }
    if (timeA > timeB) {
      return -1;
    }
    return 0;
  };
  return values.sort(compare);
}
```
But sadly I get this errors:

> No overload matches this call.
>   Overload 1 of 5, '(value: string | number | Date): Date', gave the following error.
>     Argument of type 'T[K]' is not assignable to parameter of type 'string | number | Date'.
>       Type 'T[keyof T]' is not assignable to type 'string | number | Date'.
>         Type 'T[string] | T[number] | T[symbol]' is not assignable to type 'string | number | Date'.
>           Type 'T[string]' is not assignable to type 'string | number | Date'.
>             Type 'T[string]' is not assignable to type 'Date'.
>               Type 'T[keyof T]' is not assignable to type 'Date'.
>                 Type 'T[K]' is not assignable to type 'Date'.

**Comment by bramkragten** [comment]
The helper function can be done in a new PR.

**Pull request merged**